### PR TITLE
Add message relay policies

### PR DESCRIPTION
**Pull request by thomash-acinq**
When activating onion messages, you can chose to never relay messages, relay only through existing channels or relay everything.

### CODE REVIEW

**Comment by thomash-acinq** [comment]
Replaces  #2089

**Review by t-bast — COMMENTED** [review]
I like this one better than #2089, what do you think? It doesn't need too much code, is well-contained in `MessageRelay` and the state machine is very easy to read and check.

I put a couple of nits, overall looks good to me.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 1]
```diff
@@ -249,20 +252,25 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
         d.channels.values.toSet[ActorRef].foreach(_ ! INPUT_DISCONNECTED) // we deduplicate with toSet because there might be two entries per channel (tmp id and final id)
         gotoConnected(connectionReady, d.channels)
 
-      case Event(msg: OnionMessage, _: ConnectedData) =>
+      case Event(msg: OnionMessage, d: ConnectedData) =>
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala]: Done

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala [hunk 2]
```diff
@@ -249,20 +252,25 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
...
+            case OnionMessages.SendMessage(nextNodeId, message) =>
+              switchboard ! RelayMessage(remoteNodeId, nextNodeId, message, nodeParams.onionMessageRelayPolicy)
             case received: OnionMessages.ReceiveMessage =>
               log.info(s"received message from ${remoteNodeId.value.toHex}: $received")
               context.system.eventStream.publish(received)
           }
         }
         stay()
 
+      case Event(RelayOnionMessage(msg, replyTo), d: ConnectedData) =>
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:269]: Done

File: eclair-core/src/main/scala/fr/acinq/eclair/io/MessageRelay.scala [hunk 3]
```diff
@@ -21,35 +21,74 @@ import akka.actor.typed.scaladsl.Behaviors
...
   sealed trait Command
-  case class RelayMessage(switchboard: ActorRef, nextNodeId: PublicKey, msg: OnionMessage, replyTo: typed.ActorRef[Status]) extends Command
+  case class RelayMessage(switchboard: ActorRef, prevNodeId: PublicKey, nextNodeId: PublicKey, msg: OnionMessage, policy: RelayPolicy, replyTo: typed.ActorRef[Status]) extends Command
+  case class WrappedPeerInfo(peerInfo: PeerInfoResponse) extends Command
   case class WrappedConnectionResult(result: PeerConnection.ConnectionResult) extends Command
 
   sealed trait Status
   case object Success extends Status
-  case class Failure(failure: PeerConnection.ConnectionResult.Failure) extends Status
+  case class AgainstPolicy(policy: RelayPolicy) extends Status
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/MessageRelay.scala]: Done

File: eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala [hunk 4]
```diff
@@ -51,37 +55,85 @@ class MessageRelaySpec extends ScalaTestWithActorTestKit(ConfigFactory.load("app
...
     val (_, message) = OnionMessages.buildMessage(randomKey(), randomKey(), Seq(IntermediateNode(aliceId)), Left(Recipient(bobId, None)), Nil)
-    relay ! MessageRelay.RelayMessage(switchboard.ref.toClassic, bobId, message, probe.ref)
+    relay ! RelayMessage(switchboard.ref, randomKey().publicKey, bobId, message, RelayAll, probe.ref)
 
-    val connectToNextPeer = switchboard.expectMessageType[Peer.Connect]
+    val connectToNextPeer = switchboard.expectMsgType[Peer.Connect]
     assert(connectToNextPeer.nodeId === bobId)
-    connectToNextPeer.replyTo ! PeerConnection.ConnectionResult.Connected(peerConnection.ref.toClassic)
-    peerConnection.expectMessage(message)
-    probe.expectMessage(MessageRelay.Success)
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala:60]: Done

File: eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala [hunk 5]
```diff
@@ -51,37 +55,85 @@ class MessageRelaySpec extends ScalaTestWithActorTestKit(ConfigFactory.load("app
...
+
+    val (_, message) = OnionMessages.buildMessage(randomKey(), randomKey(), Seq(IntermediateNode(aliceId)), Left(Recipient(bobId, None)), Nil)
+    val previousNodeId = randomKey().publicKey
+    relay ! RelayMessage(switchboard.ref, previousNodeId, bobId, message, RelayChannelsOnly, probe.ref)
+
+    val getPeerInfo = switchboard.expectMsgType[GetPeerInfo]
+    assert(getPeerInfo.remoteNodeId === previousNodeId)
+    getPeerInfo.replyTo ! PeerInfo(TestProbe()(system.classicSystem).ref, previousNodeId, Peer.CONNECTED, None, 0)
+
+    probe.expectMessage(AgainstPolicy(RelayChannelsOnly))
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala:103]: Done

File: eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala [hunk 6]
```diff
@@ -51,37 +55,85 @@ class MessageRelaySpec extends ScalaTestWithActorTestKit(ConfigFactory.load("app
...
+
+    val getPeerInfo1 = switchboard.expectMsgType[GetPeerInfo]
+    assert(getPeerInfo1.remoteNodeId === previousNodeId)
+    getPeerInfo1.replyTo ! PeerInfo(TestProbe()(system.classicSystem).ref, previousNodeId, Peer.CONNECTED, None, 1)
+
+    val getPeerInfo2 = switchboard.expectMsgType[GetPeerInfo]
+    assert(getPeerInfo2.remoteNodeId === bobId)
+    getPeerInfo2.replyTo ! PeerNotFound(bobId)
+
+    probe.expectMessage(AgainstPolicy(RelayChannelsOnly))
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala:122]: Done

File: eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala [hunk 7]
```diff
@@ -51,37 +55,85 @@ class MessageRelaySpec extends ScalaTestWithActorTestKit(ConfigFactory.load("app
...
+  }
+
+  test("channels on both ends") { f =>
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala:126]: Done

File: eclair-core/src/test/scala/fr/acinq/eclair/integration/MessageIntegrationSpec.scala [hunk 8]
```diff
@@ -114,13 +188,41 @@ class MessageIntegrationSpec extends IntegrationSpec {
...
+    assert(probe.expectMsgType[SendOnionMessageResponse].sent)
+
+    val r = eventListener.expectMsgType[OnionMessages.ReceiveMessage]
+    assert(r.pathId === Some(hex"2222"))
+    assert(r.finalPayload.records.unknown.toSet === Set(GenericTlv(UInt64(113), hex"010203"), GenericTlv(UInt64(117), hex"0102")))
+  }
+
+  test("channel relay with no-relay") {
+    val alice = new EclairImpl(nodes("A"))
+    connect(nodes("F"), nodes("C"))
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/integration/MessageIntegrationSpec.scala]: Done

**Review by t-bast — APPROVED** [review]
LGTM :+1:

**Pull request merged**